design/process-parameters.md

```diff
@@ -0,0 +1,65 @@
+# Process parameters
+
+We need a way to take queue-time inputs, giving pipeline authors the right amount of control / expressivity.
+
+Scenarios:
+- Pipeline users can supply different values to tools and tasks at queue time
+- Pipeline authors can control the types, ranges allowed, and defaults for queue-time parameters
+- Pipeline authors can allow flexible, queue-time control over stages/jobs to run, including adding/removing matrix legs
+- The pipeline itself can be altered at queue time: pools, agentspecs, service connections, environments, and more
+
+## Schema
+
+_Note: we need to rationalize this with the existing `parameters` syntax._
+
+```yaml
+parameters:
+# this is a sequence, so the block may be repeated as many times as needed
+- name: string          # name of the parameter; required
```

Comment:
Why would we make this significantly different than the schema we have for Task inputs?  It seems the general constraints are the same.

Reply:
Great question. I used the task schema as guidance. I felt like we've learned a lot since then, plus, the expected number of YAML authors at least is an order of magnitude higher than the number of task authors. Where it made sense, I shortened things or matched YAML precedent, since the bulk of users will not be familiar with task.json.

- **name**: matches
- **displayName**: matched our YAML model instead of task.json's _label_
- **type** / **displayHint**: this is the major departure, and its reasoning is discussed lower in the spec. "Radio" and "pickList" aren't types, they're UI widgets, and I wanted to separate them from the data type.
- **default**: in task.json it's defaultValue; seemed verbose 
- **range**: the precedent in task.json is data binding. By making it an expression-ish field, we can add that later but start by supporting simple things like enums and integer ranges.

Missing from YAML, present in task.json:
- _required_: largely meaningless in task.json, so I didn't bring it forward
- _groupName_ and _visibleRule_: didn't want to bring forward this complexity, we can add it later if required
- _properties_ and _options_: unclear what these would be needed for in YAML

Of the above, the interesting ones to discuss are **range** and **type**/**displayHint**. I'd like to get your thoughts on those specifically.

Reply:
(need to add a value for "allowed at queue time")

Reply:
(add "object" as a type - also consider if "job"/"step"/etc should be well-known)

Reply:
`range` is a little weird, find a different way